New features and improvements for version 244

karinlackrif opened the pull request:
Main changes:
* New text annotation feature.
* Edit screen for file name and description.
* Add support for offline mode.
* UI changes and improvements.

karinlackrif commented:
Hey @khawkins and @maliroteh-sf, can you please review this PR?

maliroteh-sf reviewed (commented):
Your team keeps creating PRs that are very large.... In the past we've already given feedback that you should try to break up your large PR into smaller PRs so that they can be reviewed more effectively and in a timely manner. Please consider this for your next contributions. For example you can have a PR containing cosmetics changes to html/css, another one containing logic/js changes, etc.

I glanced at this PR and added some comments for your consideration. Additionally you should perform the following:


1. You have both `package-lock.json` and `yarn.lock` files.... it is not a good idea to mix `yarn` and `npm` in one project. You should stick to one of these only so consider removing `yarn.lock` from your PR.

2. You should update your `package.json` file and change `"lint": "eslint **/{aura,lwc}/**"` to `"lint": "eslint **/lwc/**"`..... then execute `npm run lint` to catch and address linting errors. You have a lot of errors around the usage of `var`.... you should avoid using `var` and instead convert them to `let` instead. After addressing the linting issues make sure you manually verify that your component works as expected with the new changes.

3. You need to fix your tests. If you execute `npm run test` you'll see that you have failing tests.

karinlackrif commented:
Thanks @maliroteh-sf!
Unfortunately I can't split it into smaller PRs, because we don't want the customers to have access to the new features before the GA date.
1. Thanks, I removed `yarn.lock`.
2. All the errors from `eslint` are coming from the 3rd party library I'm using - `cropper.esm.js`, so I added this file to the `.eslintignore` file.
3. Fixed, thanks.

karinlackrif commented on UploadAndAnnotateImages/force-app/main/default/lwc/darkInput/darkInput.js:
```diff
@@ -0,0 +1,93 @@
+  @api
+  required;
+
+  @api
+  multiline;
+
+  // You must specify "multiline=true" when changing the number of rows
+  @api
+  rows = 1;
```
Yes I also thought about it when I wrote this code, but eventually I decided to keep it like this because if I add these setters the end result of the following code is unclear:
```
<c-dark-input multiline=false rows=3></c-dark-input>
```
Which setter will win? The `multiline` setter will set `rows=1`? Or the `rows` setter will set `multiline=true`?
So I decided that it's better to be explicit here.

karinlackrif commented on UploadAndAnnotateImages/force-app/main/default/lwc/imageAnnotate/__tests__/imageAnnotate.test.js:
```diff
@@ -199,7 +202,9 @@ describe("c-image-annotate", () => {
     );
     removeButton.click();
-    await Promise.resolve();
+    await Promise.resolve(); // Wait for the click
+    await Promise.resolve(); // Wait for LightningConfirm.open
+    await Promise.resolve(); // Wait for the event to be dispatched
```
That's how we can wait for async DOM updates in LWC tests:
https://developer.salesforce.com/docs/component-library/documentation/en/lwc/lwc.unit_testing_using_jest_create_tests#:~:text=Test%20Asynchronous%20DOM%20Updates

karinlackrif commented on UploadAndAnnotateImages/force-app/main/default/lwc/imageAnnotate/imageAnnotate.js:
```diff
@@ -111,11 +145,19 @@ export default class ImageAnnotate extends LightningElement {
   readContentHeight() {
-    const content = this.template.querySelector('[data-id="content"]');
-    this.contentHeight = content.offsetHeight;
+    // const content = this.template.querySelector('[data-id="content"]');
+    // this.contentHeight = content.offsetHeight;
```
Done, thanks!

karinlackrif commented on UploadAndAnnotateImages/force-app/main/default/lwc/imageCapture/imageCapture.js:
```diff
@@ -97,33 +98,38 @@ export default class ImageCapture extends LightningElement {
-        blob = file;
-      }
+    try {
+      for (let i = 0; i < numFiles; i++) {
+        let file = files[i];
```
Done, thanks!

karinlackrif commented on UploadAndAnnotateImages/force-app/main/default/lwc/imageInfoEditor/imageInfoEditor.html:
```diff
@@ -0,0 +1,67 @@
+              if:false={showBigButton}
+              data-id="small-button-container"
+              class="small-edit-button"
+            >
```
I used `prettier` to format all the files except JS files (and `eslint` for the JS files), so it was done automatically.

karinlackrif commented on UploadAndAnnotateImages/force-app/main/default/lwc/imageCapture/imageCapture.js-meta.xml:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8"?>
+<?xml version="1.0" encoding="UTF-8" ?>
 <LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
     <apiVersion>55.0</apiVersion>
```
Done, I updated all components to use `apiVersion = 57.0`.

maliroteh-sf reviewed (commented):
I did a fresh clone of your repo+branch and did `npm install && npm run test` and there are still test failures. 2 of your test suites fail to run so while you get 60 of 60 tests passed, I believe there are more than 60 test cases but because 2 test suites fail to even run you don't get a proper report of total number of test cases and how many failed.

Here's an example of error output I see:
```
/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/force-app/main/default/lwc/imageInfoEditor/imageInfoEditor.js:49
    return this.imageInfo.metadata.fileName;
                                   ^

TypeError: Cannot read properties of undefined (reading 'fileName')
    at ImageInfoEditor.get fileName [as fileName] (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/force-app/main/default/lwc/imageInfoEditor/imageInfoEditor.js:49:36)
    at tmpl (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/force-app/main/default/lwc/imageInfoEditor/imageInfoEditor.html:163:116)
    at /Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5404:27
    at ReactiveObserver.observe (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:713:13)
    at isUpdatingTemplate (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5356:13)
    at runWithBoundaryProtection (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:6106:5)
    at evaluateTemplate (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5349:5)
    at invokeComponentRenderMethod (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5506:41)
    at renderComponent (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5560:20)
    at rehydrate (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5855:22)
    at appendVM (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5665:3)
    at mountCustomElement (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4186:5)
    at mount (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4018:7)
    at updateStaticChildren (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4656:9)
    at patchChildren (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:3952:5)
    at patchElement (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4103:3)
    at patch (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:3987:7)
    at updateStaticChildren (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4649:11)
    at patchChildren (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:3952:5)
    at patchElement (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4103:3)
    at patch (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:3987:7)
    at updateStaticChildren (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:4649:11)
    at patchChildren (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:3952:5)
    at /Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5876:9
    at runWithBoundaryProtection (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:6106:5)
    at patchShadowRoot (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5871:7)
    at rehydrate (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5856:5)
    at Array.flushRehydrationQueue (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:5920:7)
    at flushCallbackQueue (/Users/maliroteh/Desktop/LWC-Mobile-Samples/UploadAndAnnotateImages/node_modules/@lwc/engine-dom/dist/engine-dom.cjs.js:579:21)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
 FAIL  force-app/main/default/lwc/imageAnnotate/__tests__/imageAnnotate.test.js
  ● Test suite failed to run

    Jest worker encountered 4 child process exceptions, exceeding retry limit

      at ChildProcessWorker.initialize (node_modules/jest-worker/build/workers/ChildProcessWorker.js:185:21)
```
